1 die or2

mattolbert opened the pull request:
hows this?

mattolbert committed:
created shut the box stuff

mattolbert committed:
Update ShutTheBox.py

commented out my old stuff to sort of start over. got to a point where it gives you the available tiles, you roll, you pick which ones to put down, and if the numbers you enter do work, they get removed and you roll again. I think.

mattolbert committed:
3-19 update

added combination function and other things. didn't break it down into new branch first, probably should have, but i'm still working on my work laptop. tomorrow would be good day to walk through installing what i need on home laptop to code there.

mattolbert committed:
added while loop

finally got that while loop working I think.

mattolbert committed:
3-20 updates

- added the "one die or two" element
- changed tiles from 12 to 9
- other small stuff

mcdaly commented on ShutTheBox.py:
```diff
@@ -0,0 +1,64 @@
+    break 
+
+else:
+  print("You won! You shut the box!") #i did shut the box once but this didn't appear so I need to find a way to troubleshoot that
```
Interesting. As a great Shut-the-box player, I was able to hit this!
One thought on how to troubleshoot this is to make so the list of tiles to knock down only starts with a couple numbers instead of 1-9 so it is much easier to win 
`nums = list(range(1, 4))`

mattolbert replied:
yeah this is working fine.

mcdaly commented on ShutTheBox.py:
```diff
@@ -0,0 +1,64 @@
+      die1 = random.randint(1, 6)
+      die2 = random.randint(1, 6)
+      rollvalue = die1 + die2
+  else: #i understand that this is not super elegant, I do want to try to simplify it, prob no need to have the same block twice in a row here
```
Sometimes, it needs to be done...

But a better solution would be to create a function that is called from these 3 different spots such as the following:

```
def roll_die(num_dice):
    if num_dice == 1:
        return random.randint(1, 6)
    elif num_dice == 2:
      die1 = random.randint(1, 6)
      die2 = random.randint(1, 6)
      return die1 + die2
```

mcdaly commented on ShutTheBox.py:
```diff
@@ -0,0 +1,64 @@
+    rollvalue = die1 + die2
+    
+  
+  if rollvalue == 8 or rollvalue == 11:
```
Update this to be `rollvalue in [8, 11]`
Using the list notation simplifies the code and makes it easier for adding additional numbers in the future (if needed)

mcdaly commented on ShutTheBox.py:
```diff
@@ -0,0 +1,64 @@
+    
+  
+  if rollvalue == 8 or rollvalue == 11:
+    print("You rolled an " + str(rollvalue))
```
Can you update these two lines to f-string format? Good luck!

mattolbert replied:
I'm guessing this would just be to re-write this print line as
print(f"You rolled an {rollvalue}")
no? if so, what's the advantage of this vs. what I wrote initially? if i'm misunderstanding the note here, let me know

mcdaly replied:
No advantage in this case other than getting familiar with f-string which will be more versatile and cleaner in the future

mcdaly commented on ShutTheBox.py:
```diff
@@ -0,0 +1,64 @@
+      for i in choicelist:
+        if choicemath == rollvalue and int(i) in nums:
+          nums.remove(int(i))
+          if_valid = True
```
This logic is very close but doesn't work in all cases. Essentially it breaks if the user provides more than 1 number where one number is an available number and the other is a number that has already been knocked down
 
Example:
Tiles = [1, 2, 3, 6]
rollvalue = 7
choicelist = [3, 4] (entered by me)

Expected result is that it asks me to pick the numbers again and no tiles are removed.
Unfortunately, it removes the 3, prints out "Try again" for the 4 but still proceeds to roll again and move onto the next turn

mattolbert replied:
this is what i'm currently working on that I screwed up, FYI

mcdaly commented on ShutTheBox.py:
```diff
@@ -0,0 +1,64 @@
+  return False
+
+while nums:
+  if 7 not in nums and 8 not in nums and 9 not in nums:
```
Hmmm... there is a better way to do this instead of writing out each statement separately here.
Are you familiar with list comprehension? If so, try doing this with a binary check for each [7, 8, 9] value in a list comprehensions and then use the `all()` function to wrap it.
There may or may not be enough detail here for you to do this so try it out and let me know if you get stuck

mattolbert replied:
forgot about this comment but I'll plan to get back to it

mattolbert replied:
honestly I'm still not really getting this, but I'd like to. can you explain a bit more? list comprehension is something I'm still struggling to get my head around.

mcdaly replied:
You can think of list comprehension as a way to perform a simple for loop in a single line.
Example for adding "pancake" to the end of each string
```
fruits = ['banana', 'strawberry', 'apple']
pancake_types = []
for fruit in fruits:
    pancakes_types.append(fruit + " pancakes")

# with list comprehension
pancake_types_v2 = [fruit + " pancakes" for fruit in fruits]
```

Both result in pancake_types = ['banana pancakes', 'strawberry pancakes', 'apple pancakes'] but the list comprehension is much cleaner, shorter, and doesn't require initializing the list

Lastly, it should also be noted that list comprehension is faster than spelling out the for loop following by the steps.

mcdaly replied:
On top of manipulating data within the list comprehension, you can run almost any type of function code here such as mathematical operations, functions, and comparisons.
For what I am asking out of you, the comparison would be the important part to happen for each number (7, 8, 9) of interest.

To summarize how to code up the list comprehension:
`new_variable = [<code to execute in for loop> for <cur_value> in <list_of_values>]`

mcdaly reviewed (changes requested):
Overall, looks good! Coupe minor code cleanup and edge cases that should be addressed

mattolbert committed:
added roll_die function

added roll_die function to simplify the "1 die or 2 dice" rolls, used Daly's bit of code that he suggested but i could have whipped that together on my own too

mattolbert committed:
fixed more stuff but broke the big loop

mattolbert committed:
last update of the day 3/22

i need to figure out a better system and naming convention for this but hey at least i'm not copypasting from my work comp anymore.

mcdaly commented on ShutTheBox.py:
```diff
@@ -12,19 +12,23 @@ def does_it_add_up(nums, target_num):
-      die1 = random.randint(1, 6)
-      die2 = random.randint(1, 6)
-      return die1 + die2
+  if num_dice == '1':
```
Why the change to the strings? These should stay as ints unless there is a good reason for them to be strings

mcdaly commented on ShutTheBox.py:
```diff
@@ -12,19 +12,23 @@ def does_it_add_up(nums, target_num):
+    rollval = die1 + die2
+  return rollval
+
+def has_duplicates(seq):#i built this because I accidentally typed "1, 1" in as user input and it did all sort of weirdness, I think this fixed it
```
Nice!

mcdaly commented on ShutTheBox.py:
```diff
@@ -12,19 +12,23 @@ def does_it_add_up(nums, target_num):
     rollvalue = roll_die(dice_choice)
   else:
-    rollvalue = roll_die(2)
+    rollvalue = roll_die('2')
```
back to int!

mcdaly commented on ShutTheBox.py:
```diff
@@ -36,19 +40,20 @@ def roll_die(num_dice):
-      choicemath = 0
-      for i in choicelist:
-        choicemath += int(i)
+      intlist = [int(i) for i in choicelist]#can this and the line above it all happen at once?
```
Yes it can! Just put `int(i.strip())` so the choicelist is just a list of ints. This does require the user to only enter numbers (no letters or empty strings) but I think that is a fine assumption for this code.

mattolbert replied:
thanks!

mcdaly commented on ShutTheBox.py:
```diff
@@ -36,19 +40,20 @@ def roll_die(num_dice):
-      for i in choicelist:
-        choicemath += int(i)
+      intlist = [int(i) for i in choicelist]#can this and the line above it all happen at once?
+      choicemath = sum(intlist)
```
Nice!

mcdaly commented on ShutTheBox.py:
```diff
@@ -36,19 +40,20 @@ def roll_die(num_dice):
+      if choicemath == rollvalue:
+        if has_duplicates(intlist) == True:
+          print("You can't put the same number down twice")
+          break
```
I am seeing when I enter a double number that no numbers are put down but the dice are rolled again. This should not be the case.
I believe some confusion is around the `break` statement. The `break` statement terminates the current `for` or `while` loop being executed. In this case, this `break` statement is in the `while if_valid` so it is breaking out of the user input check and jumping to the next roll

mcdaly commented on ShutTheBox.py:
```diff
@@ -36,19 +40,20 @@ def roll_die(num_dice):
+              break       
+            nums.remove(i)
+            if_valid = True
       else:
```
Code is working well (except for the duplicate comment)!

This could be easier to follow if the print statement was next to the if statement. Here is a potential way to update this code block to simplify all the nested if statements

```   
      if choicemath != rollvalue:
        print("Your choice doesn't match the roll value")
      elif has_duplicates(intlist) == True:
        print("You can't put the same number down twice")
        break
      else:
        # Ensure all requested tiles are still up
        any_nums_down_already = any(i not in nums for i in intlist)
        if any_nums_down_already:
            print("Your number choice has already been put down")
        else:
          # Good inputs. Knock down tiles
          for i in intlist:
            nums.remove(i)
          if_valid = True
```

mattolbert replied:
this makes total sense. looking at mine, since I sort of was adding these exceptions and error messages as I went along, they ended up being sort of cobbled together. Had I thought about them all at the beginning, I mightve thought to streamline them together like this. good feedback.

mattolbert committed:
what are good names to use for individual commits like this?
